cdap-watchdog/src/main/java/io/cdap/cdap/logging/gateway/handlers/RemoteLogsFetcher.java

```diff
@@ -0,0 +1,121 @@
+import io.cdap.common.http.HttpContentConsumer;
+import io.cdap.common.http.HttpRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+/**
+ * Fetch Program logs via internal REST API calls
+ */
+public class RemoteLogsFetcher implements LogsFetcher {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteLogsFetcher.class);
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemoteLogsFetcher(RemoteClientFactory remoteClientFactory) {
+    this.remoteClient =
+      remoteClientFactory.createRemoteClient(Constants.Service.LOG_QUERY, new DefaultHttpRequestConfig(false),
+                                             Gateway.API_VERSION_3);
+  }
+
+  /**
+   * Gets the run logs of a program.
+   *
+   * @param program the program
+   * @param runId pipeline run id
+   * @param start start time of the time range of desired logs
+   * @param stop end time of the time range of desired logs
+   * @param file file for us to write the log into
+   * @throws IOException              if a network error occurred
+   * @throws UnauthenticatedException if the request is not authorized successfully in the gateway
+   *                                  server
+   */
+  @Override
+  public void getProgramRunLogs(ProgramId program, String runId, long start, long stop, File file)
+    throws IOException, UnauthenticatedException, UnauthorizedException {
+
+    String path = String.format("namespaces/%s/apps/%s/%s/%s/runs/%s/logs?start=%d&stop=%d",
+                                program.getNamespaceId().getNamespace(), program.getApplication(),
+                                program.getType().getCategoryName(), program.getProgram(), runId, start, stop);
+    FileChannel channel = new FileOutputStream(file, false).getChannel();
+    URL url = remoteClient.resolve(path);
+    execute(url, channel, file);
+  }
+
+  /**
+   * Gets the logs of a program.
+   *
+   * @param componentId component id
+   * @param serviceId service id
+   * @param start start time of the time range of desired logs
+   * @param stop end time of the time range of desired logs
+   * @param file file for us to write the log into
+   * @throws IOException              if a network error occurred
+   * @throws UnauthenticatedException if the request is not authorized successfully in the gateway
+   *                                  server
+   */
+  @Override
+  public void getSystemServiceLog(String componentId, String serviceId, long start, long stop, File file)
+    throws IOException, UnauthenticatedException, UnauthorizedException {
+    String path = String.format("system/%s/%s/logs?start=%d&stop=%d", componentId, serviceId, start, stop);
+    FileChannel channel = new FileOutputStream(file, false).getChannel();
```

Comment:
Please use tryWithResource to ensure file is closed even if there are issues with connecting to URI